In [80]:
import pandas as pd

# We will be cleaning customer call records data so that it's easier for those making calls. 


# Getting a sense of the data

In [84]:
df = pd.read_excel("Customer Call List.xlsx")

In [86]:
df.head(21)

CustomerID First_Name    Last_Name  Phone_Number  \
0         1001      Frodo      Baggins  123-545-5421   
1         1002       Abed        Nadir  123/643/9775   
2         1003     Walter       /White    7066950392   
3         1004     Dwight      Schrute  123-543-2345   
4         1005        Jon         Snow  876|678|3469   
5         1006        Ron      Swanson  304-762-2467   
6         1007       Jeff       Winger           NaN   
7         1008   Sherlock       Holmes  876|678|3469   
8         1009    Gandalf          NaN           N/a   
9         1010      Peter       Parker  123-545-5421   
10        1011    Samwise       Gamgee           NaN   
11        1012      Harry    ...Potter    7066950392   
12        1013        Don       Draper  123-543-2345   
13        1014     Leslie        Knope  876|678|3469   
14        1015       Toby  Flenderson_  304-762-2467   
15        1016        Ron      Weasley  123-545-5421   
16        1017   Michael         Scott  123/643/9775   
17        1018      Clark         Kent    7066950392   
18        1019      Creed       Braton           N/a   
19        1020     Anakin    Skywalker  876|678|3469   
20        1020     Anakin    Skywalker  876|678|3469   

                                  Address Paying Customer Do_Not_Contact  \
0                   123 Shire Lane, Shire             Yes             No   
1                     93 West Main Street              No            Yes   
2                      298 Drugs Driveway               N            NaN   
3   980 Paper Avenue, Pennsylvania, 18503             Yes              Y   
4                        123 Dragons Road               Y             No   
5                        768 City Parkway             Yes            Yes   
6                       1209 South Street              No             No   
7                           98 Clue Drive               N             No   
8                        123 Middle Earth             Yes            NaN   
9              25th Main Street, New York             Yes             No   
10                  612 Shire Lane, Shire             Yes             No   
11                   2394 Hogwarts Avenue               Y            NaN   
12                       2039 Main Street             Yes              N   
13                       343 City Parkway             Yes             No   
14                          214 HR Avenue               N             No   
15                   2395 Hogwarts Avenue              No              N   
16         121 Paper Avenue, Pennsylvania             Yes             No   
17                        3498 Super Lane               Y            NaN   
18                                    N/a             N/a            Yes   
19            910 Tatooine Road, Tatooine             Yes              N   
20            910 Tatooine Road, Tatooine             Yes              N   

    Not_Useful_Column  
0                True  
1               False  
2                True  
3                True  
4                True  
5                True  
6               False  
7               False  
8               False  
9                True  
10               True  
11               True  
12              False  
13              False  
14              False  
15              False  
16              False  
17               True  
18               True  
19               True  
20               True

# Drop duplicates first

In [89]:
df = df.drop_duplicates()

In [91]:
df

CustomerID First_Name    Last_Name  Phone_Number  \
0         1001      Frodo      Baggins  123-545-5421   
1         1002       Abed        Nadir  123/643/9775   
2         1003     Walter       /White    7066950392   
3         1004     Dwight      Schrute  123-543-2345   
4         1005        Jon         Snow  876|678|3469   
5         1006        Ron      Swanson  304-762-2467   
6         1007       Jeff       Winger           NaN   
7         1008   Sherlock       Holmes  876|678|3469   
8         1009    Gandalf          NaN           N/a   
9         1010      Peter       Parker  123-545-5421   
10        1011    Samwise       Gamgee           NaN   
11        1012      Harry    ...Potter    7066950392   
12        1013        Don       Draper  123-543-2345   
13        1014     Leslie        Knope  876|678|3469   
14        1015       Toby  Flenderson_  304-762-2467   
15        1016        Ron      Weasley  123-545-5421   
16        1017   Michael         Scott  123/643/9775   
17        1018      Clark         Kent    7066950392   
18        1019      Creed       Braton           N/a   
19        1020     Anakin    Skywalker  876|678|3469   

                                  Address Paying Customer Do_Not_Contact  \
0                   123 Shire Lane, Shire             Yes             No   
1                     93 West Main Street              No            Yes   
2                      298 Drugs Driveway               N            NaN   
3   980 Paper Avenue, Pennsylvania, 18503             Yes              Y   
4                        123 Dragons Road               Y             No   
5                        768 City Parkway             Yes            Yes   
6                       1209 South Street              No             No   
7                           98 Clue Drive               N             No   
8                        123 Middle Earth             Yes            NaN   
9              25th Main Street, New York             Yes             No   
10                  612 Shire Lane, Shire             Yes             No   
11                   2394 Hogwarts Avenue               Y            NaN   
12                       2039 Main Street             Yes              N   
13                       343 City Parkway             Yes             No   
14                          214 HR Avenue               N             No   
15                   2395 Hogwarts Avenue              No              N   
16         121 Paper Avenue, Pennsylvania             Yes             No   
17                        3498 Super Lane               Y            NaN   
18                                    N/a             N/a            Yes   
19            910 Tatooine Road, Tatooine             Yes              N   

    Not_Useful_Column  
0                True  
1               False  
2                True  
3                True  
4                True  
5                True  
6               False  
7               False  
8               False  
9                True  
10               True  
11               True  
12              False  
13              False  
14              False  
15              False  
16              False  
17               True  
18               True  
19               True

# Drop not_useful_column

In [94]:
df = df.drop(columns = ["Not_Useful_Column"])

df

CustomerID First_Name    Last_Name  Phone_Number  \
0         1001      Frodo      Baggins  123-545-5421   
1         1002       Abed        Nadir  123/643/9775   
2         1003     Walter       /White    7066950392   
3         1004     Dwight      Schrute  123-543-2345   
4         1005        Jon         Snow  876|678|3469   
5         1006        Ron      Swanson  304-762-2467   
6         1007       Jeff       Winger           NaN   
7         1008   Sherlock       Holmes  876|678|3469   
8         1009    Gandalf          NaN           N/a   
9         1010      Peter       Parker  123-545-5421   
10        1011    Samwise       Gamgee           NaN   
11        1012      Harry    ...Potter    7066950392   
12        1013        Don       Draper  123-543-2345   
13        1014     Leslie        Knope  876|678|3469   
14        1015       Toby  Flenderson_  304-762-2467   
15        1016        Ron      Weasley  123-545-5421   
16        1017   Michael         Scott  123/643/9775   
17        1018      Clark         Kent    7066950392   
18        1019      Creed       Braton           N/a   
19        1020     Anakin    Skywalker  876|678|3469   

                                  Address Paying Customer Do_Not_Contact  
0                   123 Shire Lane, Shire             Yes             No  
1                     93 West Main Street              No            Yes  
2                      298 Drugs Driveway               N            NaN  
3   980 Paper Avenue, Pennsylvania, 18503             Yes              Y  
4                        123 Dragons Road               Y             No  
5                        768 City Parkway             Yes            Yes  
6                       1209 South Street              No             No  
7                           98 Clue Drive               N             No  
8                        123 Middle Earth             Yes            NaN  
9              25th Main Street, New York             Yes             No  
10                  612 Shire Lane, Shire             Yes             No  
11                   2394 Hogwarts Avenue               Y            NaN  
12                       2039 Main Street             Yes              N  
13                       343 City Parkway             Yes             No  
14                          214 HR Avenue               N             No  
15                   2395 Hogwarts Avenue              No              N  
16         121 Paper Avenue, Pennsylvania             Yes             No  
17                        3498 Super Lane               Y            NaN  
18                                    N/a             N/a            Yes  
19            910 Tatooine Road, Tatooine             Yes              N

# Go column by column and see what needs to be cleaned/fixed

In [97]:
df["Last_Name"] = df["Last_Name"].str.lstrip("...") #apply to only this column

df

CustomerID First_Name    Last_Name  Phone_Number  \
0         1001      Frodo      Baggins  123-545-5421   
1         1002       Abed        Nadir  123/643/9775   
2         1003     Walter       /White    7066950392   
3         1004     Dwight      Schrute  123-543-2345   
4         1005        Jon         Snow  876|678|3469   
5         1006        Ron      Swanson  304-762-2467   
6         1007       Jeff       Winger           NaN   
7         1008   Sherlock       Holmes  876|678|3469   
8         1009    Gandalf          NaN           N/a   
9         1010      Peter       Parker  123-545-5421   
10        1011    Samwise       Gamgee           NaN   
11        1012      Harry       Potter    7066950392   
12        1013        Don       Draper  123-543-2345   
13        1014     Leslie        Knope  876|678|3469   
14        1015       Toby  Flenderson_  304-762-2467   
15        1016        Ron      Weasley  123-545-5421   
16        1017   Michael         Scott  123/643/9775   
17        1018      Clark         Kent    7066950392   
18        1019      Creed       Braton           N/a   
19        1020     Anakin    Skywalker  876|678|3469   

                                  Address Paying Customer Do_Not_Contact  
0                   123 Shire Lane, Shire             Yes             No  
1                     93 West Main Street              No            Yes  
2                      298 Drugs Driveway               N            NaN  
3   980 Paper Avenue, Pennsylvania, 18503             Yes              Y  
4                        123 Dragons Road               Y             No  
5                        768 City Parkway             Yes            Yes  
6                       1209 South Street              No             No  
7                           98 Clue Drive               N             No  
8                        123 Middle Earth             Yes            NaN  
9              25th Main Street, New York             Yes             No  
10                  612 Shire Lane, Shire             Yes             No  
11                   2394 Hogwarts Avenue               Y            NaN  
12                       2039 Main Street             Yes              N  
13                       343 City Parkway             Yes             No  
14                          214 HR Avenue               N             No  
15                   2395 Hogwarts Avenue              No              N  
16         121 Paper Avenue, Pennsylvania             Yes             No  
17                        3498 Super Lane               Y            NaN  
18                                    N/a             N/a            Yes  
19            910 Tatooine Road, Tatooine             Yes              N

In [99]:
df["Last_Name"] = df["Last_Name"].str.rstrip("_") #apply to only this column
df["Last_Name"] = df["Last_Name"].str.lstrip("/") #apply to only this column

df

CustomerID First_Name   Last_Name  Phone_Number  \
0         1001      Frodo     Baggins  123-545-5421   
1         1002       Abed       Nadir  123/643/9775   
2         1003     Walter       White    7066950392   
3         1004     Dwight     Schrute  123-543-2345   
4         1005        Jon        Snow  876|678|3469   
5         1006        Ron     Swanson  304-762-2467   
6         1007       Jeff      Winger           NaN   
7         1008   Sherlock      Holmes  876|678|3469   
8         1009    Gandalf         NaN           N/a   
9         1010      Peter      Parker  123-545-5421   
10        1011    Samwise      Gamgee           NaN   
11        1012      Harry      Potter    7066950392   
12        1013        Don      Draper  123-543-2345   
13        1014     Leslie       Knope  876|678|3469   
14        1015       Toby  Flenderson  304-762-2467   
15        1016        Ron     Weasley  123-545-5421   
16        1017   Michael        Scott  123/643/9775   
17        1018      Clark        Kent    7066950392   
18        1019      Creed      Braton           N/a   
19        1020     Anakin   Skywalker  876|678|3469   

                                  Address Paying Customer Do_Not_Contact  
0                   123 Shire Lane, Shire             Yes             No  
1                     93 West Main Street              No            Yes  
2                      298 Drugs Driveway               N            NaN  
3   980 Paper Avenue, Pennsylvania, 18503             Yes              Y  
4                        123 Dragons Road               Y             No  
5                        768 City Parkway             Yes            Yes  
6                       1209 South Street              No             No  
7                           98 Clue Drive               N             No  
8                        123 Middle Earth             Yes            NaN  
9              25th Main Street, New York             Yes             No  
10                  612 Shire Lane, Shire             Yes             No  
11                   2394 Hogwarts Avenue               Y            NaN  
12                       2039 Main Street             Yes              N  
13                       343 City Parkway             Yes             No  
14                          214 HR Avenue               N             No  
15                   2395 Hogwarts Avenue              No              N  
16         121 Paper Avenue, Pennsylvania             Yes             No  
17                        3498 Super Lane               Y            NaN  
18                                    N/a             N/a            Yes  
19            910 Tatooine Road, Tatooine             Yes              N

#alternatively we can do which will clean up 

df["Last_Name"].str.strip("123._/") #set of characters to remove as leading/trailing characters

# Standardising format (for phone numbers) 

In [103]:
#1st we will use regex to remove all alphanumeric characters
#next we will standardise the phone numbers

df["Phone_Number"] = df["Phone_Number"].str.replace('[^a-zA-Z0-9]','',regex=True) #we will replace all non (^) alphanumeric characters with nothing so essentially removing them 

df

CustomerID First_Name   Last_Name Phone_Number  \
0         1001      Frodo     Baggins   1235455421   
1         1002       Abed       Nadir   1236439775   
2         1003     Walter       White          NaN   
3         1004     Dwight     Schrute   1235432345   
4         1005        Jon        Snow   8766783469   
5         1006        Ron     Swanson   3047622467   
6         1007       Jeff      Winger          NaN   
7         1008   Sherlock      Holmes   8766783469   
8         1009    Gandalf         NaN           Na   
9         1010      Peter      Parker   1235455421   
10        1011    Samwise      Gamgee          NaN   
11        1012      Harry      Potter          NaN   
12        1013        Don      Draper   1235432345   
13        1014     Leslie       Knope   8766783469   
14        1015       Toby  Flenderson   3047622467   
15        1016        Ron     Weasley   1235455421   
16        1017   Michael        Scott   1236439775   
17        1018      Clark        Kent          NaN   
18        1019      Creed      Braton           Na   
19        1020     Anakin   Skywalker   8766783469   

                                  Address Paying Customer Do_Not_Contact  
0                   123 Shire Lane, Shire             Yes             No  
1                     93 West Main Street              No            Yes  
2                      298 Drugs Driveway               N            NaN  
3   980 Paper Avenue, Pennsylvania, 18503             Yes              Y  
4                        123 Dragons Road               Y             No  
5                        768 City Parkway             Yes            Yes  
6                       1209 South Street              No             No  
7                           98 Clue Drive               N             No  
8                        123 Middle Earth             Yes            NaN  
9              25th Main Street, New York             Yes             No  
10                  612 Shire Lane, Shire             Yes             No  
11                   2394 Hogwarts Avenue               Y            NaN  
12                       2039 Main Street             Yes              N  
13                       343 City Parkway             Yes             No  
14                          214 HR Avenue               N             No  
15                   2395 Hogwarts Avenue              No              N  
16         121 Paper Avenue, Pennsylvania             Yes             No  
17                        3498 Super Lane               Y            NaN  
18                                    N/a             N/a            Yes  
19            910 Tatooine Road, Tatooine             Yes              N

In [105]:
#to standardise phone numbers - 1st convert to string 

df["Phone_Number"]= df["Phone_Number"].apply(lambda x: str(x))

df

CustomerID First_Name   Last_Name Phone_Number  \
0         1001      Frodo     Baggins   1235455421   
1         1002       Abed       Nadir   1236439775   
2         1003     Walter       White          nan   
3         1004     Dwight     Schrute   1235432345   
4         1005        Jon        Snow   8766783469   
5         1006        Ron     Swanson   3047622467   
6         1007       Jeff      Winger          nan   
7         1008   Sherlock      Holmes   8766783469   
8         1009    Gandalf         NaN           Na   
9         1010      Peter      Parker   1235455421   
10        1011    Samwise      Gamgee          nan   
11        1012      Harry      Potter          nan   
12        1013        Don      Draper   1235432345   
13        1014     Leslie       Knope   8766783469   
14        1015       Toby  Flenderson   3047622467   
15        1016        Ron     Weasley   1235455421   
16        1017   Michael        Scott   1236439775   
17        1018      Clark        Kent          nan   
18        1019      Creed      Braton           Na   
19        1020     Anakin   Skywalker   8766783469   

                                  Address Paying Customer Do_Not_Contact  
0                   123 Shire Lane, Shire             Yes             No  
1                     93 West Main Street              No            Yes  
2                      298 Drugs Driveway               N            NaN  
3   980 Paper Avenue, Pennsylvania, 18503             Yes              Y  
4                        123 Dragons Road               Y             No  
5                        768 City Parkway             Yes            Yes  
6                       1209 South Street              No             No  
7                           98 Clue Drive               N             No  
8                        123 Middle Earth             Yes            NaN  
9              25th Main Street, New York             Yes             No  
10                  612 Shire Lane, Shire             Yes             No  
11                   2394 Hogwarts Avenue               Y            NaN  
12                       2039 Main Street             Yes              N  
13                       343 City Parkway             Yes             No  
14                          214 HR Avenue               N             No  
15                   2395 Hogwarts Avenue              No              N  
16         121 Paper Avenue, Pennsylvania             Yes             No  
17                        3498 Super Lane               Y            NaN  
18                                    N/a             N/a            Yes  
19            910 Tatooine Road, Tatooine             Yes              N

In [107]:
#then now standardise

df["Phone_Number"] = df["Phone_Number"].apply(lambda x:x[0:3]+'-'+x[3:6]+'-'+x[6:10])

df

CustomerID First_Name   Last_Name  Phone_Number  \
0         1001      Frodo     Baggins  123-545-5421   
1         1002       Abed       Nadir  123-643-9775   
2         1003     Walter       White         nan--   
3         1004     Dwight     Schrute  123-543-2345   
4         1005        Jon        Snow  876-678-3469   
5         1006        Ron     Swanson  304-762-2467   
6         1007       Jeff      Winger         nan--   
7         1008   Sherlock      Holmes  876-678-3469   
8         1009    Gandalf         NaN          Na--   
9         1010      Peter      Parker  123-545-5421   
10        1011    Samwise      Gamgee         nan--   
11        1012      Harry      Potter         nan--   
12        1013        Don      Draper  123-543-2345   
13        1014     Leslie       Knope  876-678-3469   
14        1015       Toby  Flenderson  304-762-2467   
15        1016        Ron     Weasley  123-545-5421   
16        1017   Michael        Scott  123-643-9775   
17        1018      Clark        Kent         nan--   
18        1019      Creed      Braton          Na--   
19        1020     Anakin   Skywalker  876-678-3469   

                                  Address Paying Customer Do_Not_Contact  
0                   123 Shire Lane, Shire             Yes             No  
1                     93 West Main Street              No            Yes  
2                      298 Drugs Driveway               N            NaN  
3   980 Paper Avenue, Pennsylvania, 18503             Yes              Y  
4                        123 Dragons Road               Y             No  
5                        768 City Parkway             Yes            Yes  
6                       1209 South Street              No             No  
7                           98 Clue Drive               N             No  
8                        123 Middle Earth             Yes            NaN  
9              25th Main Street, New York             Yes             No  
10                  612 Shire Lane, Shire             Yes             No  
11                   2394 Hogwarts Avenue               Y            NaN  
12                       2039 Main Street             Yes              N  
13                       343 City Parkway             Yes             No  
14                          214 HR Avenue               N             No  
15                   2395 Hogwarts Avenue              No              N  
16         121 Paper Avenue, Pennsylvania             Yes             No  
17                        3498 Super Lane               Y            NaN  
18                                    N/a             N/a            Yes  
19            910 Tatooine Road, Tatooine             Yes              N

In [109]:
df["Phone_Number"] = df["Phone_Number"].str.replace('nan--','')
df["Phone_Number"] = df["Phone_Number"].str.replace('Na--','')

df

CustomerID First_Name   Last_Name  Phone_Number  \
0         1001      Frodo     Baggins  123-545-5421   
1         1002       Abed       Nadir  123-643-9775   
2         1003     Walter       White                 
3         1004     Dwight     Schrute  123-543-2345   
4         1005        Jon        Snow  876-678-3469   
5         1006        Ron     Swanson  304-762-2467   
6         1007       Jeff      Winger                 
7         1008   Sherlock      Holmes  876-678-3469   
8         1009    Gandalf         NaN                 
9         1010      Peter      Parker  123-545-5421   
10        1011    Samwise      Gamgee                 
11        1012      Harry      Potter                 
12        1013        Don      Draper  123-543-2345   
13        1014     Leslie       Knope  876-678-3469   
14        1015       Toby  Flenderson  304-762-2467   
15        1016        Ron     Weasley  123-545-5421   
16        1017   Michael        Scott  123-643-9775   
17        1018      Clark        Kent                 
18        1019      Creed      Braton                 
19        1020     Anakin   Skywalker  876-678-3469   

                                  Address Paying Customer Do_Not_Contact  
0                   123 Shire Lane, Shire             Yes             No  
1                     93 West Main Street              No            Yes  
2                      298 Drugs Driveway               N            NaN  
3   980 Paper Avenue, Pennsylvania, 18503             Yes              Y  
4                        123 Dragons Road               Y             No  
5                        768 City Parkway             Yes            Yes  
6                       1209 South Street              No             No  
7                           98 Clue Drive               N             No  
8                        123 Middle Earth             Yes            NaN  
9              25th Main Street, New York             Yes             No  
10                  612 Shire Lane, Shire             Yes             No  
11                   2394 Hogwarts Avenue               Y            NaN  
12                       2039 Main Street             Yes              N  
13                       343 City Parkway             Yes             No  
14                          214 HR Avenue               N             No  
15                   2395 Hogwarts Avenue              No              N  
16         121 Paper Avenue, Pennsylvania             Yes             No  
17                        3498 Super Lane               Y            NaN  
18                                    N/a             N/a            Yes  
19            910 Tatooine Road, Tatooine             Yes              N

# Splitting into columns for address

In [112]:
df[["Street_Address","State","Zip_Code"]] =df["Address"].str.split(",",n=2,expand = True) #split at most 2 times so we get 3 outputs
#expand=True returns a dataframe instead of a series of lists

In [114]:
df

CustomerID First_Name   Last_Name  Phone_Number  \
0         1001      Frodo     Baggins  123-545-5421   
1         1002       Abed       Nadir  123-643-9775   
2         1003     Walter       White                 
3         1004     Dwight     Schrute  123-543-2345   
4         1005        Jon        Snow  876-678-3469   
5         1006        Ron     Swanson  304-762-2467   
6         1007       Jeff      Winger                 
7         1008   Sherlock      Holmes  876-678-3469   
8         1009    Gandalf         NaN                 
9         1010      Peter      Parker  123-545-5421   
10        1011    Samwise      Gamgee                 
11        1012      Harry      Potter                 
12        1013        Don      Draper  123-543-2345   
13        1014     Leslie       Knope  876-678-3469   
14        1015       Toby  Flenderson  304-762-2467   
15        1016        Ron     Weasley  123-545-5421   
16        1017   Michael        Scott  123-643-9775   
17        1018      Clark        Kent                 
18        1019      Creed      Braton                 
19        1020     Anakin   Skywalker  876-678-3469   

                                  Address Paying Customer Do_Not_Contact  \
0                   123 Shire Lane, Shire             Yes             No   
1                     93 West Main Street              No            Yes   
2                      298 Drugs Driveway               N            NaN   
3   980 Paper Avenue, Pennsylvania, 18503             Yes              Y   
4                        123 Dragons Road               Y             No   
5                        768 City Parkway             Yes            Yes   
6                       1209 South Street              No             No   
7                           98 Clue Drive               N             No   
8                        123 Middle Earth             Yes            NaN   
9              25th Main Street, New York             Yes             No   
10                  612 Shire Lane, Shire             Yes             No   
11                   2394 Hogwarts Avenue               Y            NaN   
12                       2039 Main Street             Yes              N   
13                       343 City Parkway             Yes             No   
14                          214 HR Avenue               N             No   
15                   2395 Hogwarts Avenue              No              N   
16         121 Paper Avenue, Pennsylvania             Yes             No   
17                        3498 Super Lane               Y            NaN   
18                                    N/a             N/a            Yes   
19            910 Tatooine Road, Tatooine             Yes              N   

          Street_Address          State Zip_Code  
0         123 Shire Lane          Shire     None  
1    93 West Main Street           None     None  
2     298 Drugs Driveway           None     None  
3       980 Paper Avenue   Pennsylvania    18503  
4       123 Dragons Road           None     None  
5       768 City Parkway           None     None  
6      1209 South Street           None     None  
7          98 Clue Drive           None     None  
8       123 Middle Earth           None     None  
9       25th Main Street       New York     None  
10        612 Shire Lane          Shire     None  
11  2394 Hogwarts Avenue           None     None  
12      2039 Main Street           None     None  
13      343 City Parkway           None     None  
14         214 HR Avenue           None     None  
15  2395 Hogwarts Avenue           None     None  
16      121 Paper Avenue   Pennsylvania     None  
17       3498 Super Lane           None     None  
18                   N/a           None     None  
19     910 Tatooine Road       Tatooine     None

# Standardising for Do_Not_Contact and Paying Customer

In [117]:
df["Paying Customer"] = df["Paying Customer"].str.replace('Yes','Y')

df["Paying Customer"] = df["Paying Customer"].str.replace('No','N')

df

CustomerID First_Name   Last_Name  Phone_Number  \
0         1001      Frodo     Baggins  123-545-5421   
1         1002       Abed       Nadir  123-643-9775   
2         1003     Walter       White                 
3         1004     Dwight     Schrute  123-543-2345   
4         1005        Jon        Snow  876-678-3469   
5         1006        Ron     Swanson  304-762-2467   
6         1007       Jeff      Winger                 
7         1008   Sherlock      Holmes  876-678-3469   
8         1009    Gandalf         NaN                 
9         1010      Peter      Parker  123-545-5421   
10        1011    Samwise      Gamgee                 
11        1012      Harry      Potter                 
12        1013        Don      Draper  123-543-2345   
13        1014     Leslie       Knope  876-678-3469   
14        1015       Toby  Flenderson  304-762-2467   
15        1016        Ron     Weasley  123-545-5421   
16        1017   Michael        Scott  123-643-9775   
17        1018      Clark        Kent                 
18        1019      Creed      Braton                 
19        1020     Anakin   Skywalker  876-678-3469   

                                  Address Paying Customer Do_Not_Contact  \
0                   123 Shire Lane, Shire               Y             No   
1                     93 West Main Street               N            Yes   
2                      298 Drugs Driveway               N            NaN   
3   980 Paper Avenue, Pennsylvania, 18503               Y              Y   
4                        123 Dragons Road               Y             No   
5                        768 City Parkway               Y            Yes   
6                       1209 South Street               N             No   
7                           98 Clue Drive               N             No   
8                        123 Middle Earth               Y            NaN   
9              25th Main Street, New York               Y             No   
10                  612 Shire Lane, Shire               Y             No   
11                   2394 Hogwarts Avenue               Y            NaN   
12                       2039 Main Street               Y              N   
13                       343 City Parkway               Y             No   
14                          214 HR Avenue               N             No   
15                   2395 Hogwarts Avenue               N              N   
16         121 Paper Avenue, Pennsylvania               Y             No   
17                        3498 Super Lane               Y            NaN   
18                                    N/a             N/a            Yes   
19            910 Tatooine Road, Tatooine               Y              N   

          Street_Address          State Zip_Code  
0         123 Shire Lane          Shire     None  
1    93 West Main Street           None     None  
2     298 Drugs Driveway           None     None  
3       980 Paper Avenue   Pennsylvania    18503  
4       123 Dragons Road           None     None  
5       768 City Parkway           None     None  
6      1209 South Street           None     None  
7          98 Clue Drive           None     None  
8       123 Middle Earth           None     None  
9       25th Main Street       New York     None  
10        612 Shire Lane          Shire     None  
11  2394 Hogwarts Avenue           None     None  
12      2039 Main Street           None     None  
13      343 City Parkway           None     None  
14         214 HR Avenue           None     None  
15  2395 Hogwarts Avenue           None     None  
16      121 Paper Avenue   Pennsylvania     None  
17       3498 Super Lane           None     None  
18                   N/a           None     None  
19     910 Tatooine Road       Tatooine     None

In [119]:
df["Do_Not_Contact"] = df["Do_Not_Contact"].str.replace('No','N')
df["Do_Not_Contact"] = df["Do_Not_Contact"].str.replace('Yes','Y')


# Dealing with empty values

In [122]:
df = df.fillna('')

df

CustomerID First_Name   Last_Name  Phone_Number  \
0         1001      Frodo     Baggins  123-545-5421   
1         1002       Abed       Nadir  123-643-9775   
2         1003     Walter       White                 
3         1004     Dwight     Schrute  123-543-2345   
4         1005        Jon        Snow  876-678-3469   
5         1006        Ron     Swanson  304-762-2467   
6         1007       Jeff      Winger                 
7         1008   Sherlock      Holmes  876-678-3469   
8         1009    Gandalf                             
9         1010      Peter      Parker  123-545-5421   
10        1011    Samwise      Gamgee                 
11        1012      Harry      Potter                 
12        1013        Don      Draper  123-543-2345   
13        1014     Leslie       Knope  876-678-3469   
14        1015       Toby  Flenderson  304-762-2467   
15        1016        Ron     Weasley  123-545-5421   
16        1017   Michael        Scott  123-643-9775   
17        1018      Clark        Kent                 
18        1019      Creed      Braton                 
19        1020     Anakin   Skywalker  876-678-3469   

                                  Address Paying Customer Do_Not_Contact  \
0                   123 Shire Lane, Shire               Y              N   
1                     93 West Main Street               N              Y   
2                      298 Drugs Driveway               N                  
3   980 Paper Avenue, Pennsylvania, 18503               Y              Y   
4                        123 Dragons Road               Y              N   
5                        768 City Parkway               Y              Y   
6                       1209 South Street               N              N   
7                           98 Clue Drive               N              N   
8                        123 Middle Earth               Y                  
9              25th Main Street, New York               Y              N   
10                  612 Shire Lane, Shire               Y              N   
11                   2394 Hogwarts Avenue               Y                  
12                       2039 Main Street               Y              N   
13                       343 City Parkway               Y              N   
14                          214 HR Avenue               N              N   
15                   2395 Hogwarts Avenue               N              N   
16         121 Paper Avenue, Pennsylvania               Y              N   
17                        3498 Super Lane               Y                  
18                                    N/a             N/a              Y   
19            910 Tatooine Road, Tatooine               Y              N   

          Street_Address          State Zip_Code  
0         123 Shire Lane          Shire           
1    93 West Main Street                          
2     298 Drugs Driveway                          
3       980 Paper Avenue   Pennsylvania    18503  
4       123 Dragons Road                          
5       768 City Parkway                          
6      1209 South Street                          
7          98 Clue Drive                          
8       123 Middle Earth                          
9       25th Main Street       New York           
10        612 Shire Lane          Shire           
11  2394 Hogwarts Avenue                          
12      2039 Main Street                          
13      343 City Parkway                          
14         214 HR Avenue                          
15  2395 Hogwarts Avenue                          
16      121 Paper Avenue   Pennsylvania           
17       3498 Super Lane                          
18                   N/a                          
19     910 Tatooine Road       Tatooine

In [126]:
df["Address"] = df["Address"].str.replace("N/a",'')

df

CustomerID First_Name   Last_Name  Phone_Number  \
0         1001      Frodo     Baggins  123-545-5421   
1         1002       Abed       Nadir  123-643-9775   
2         1003     Walter       White                 
3         1004     Dwight     Schrute  123-543-2345   
4         1005        Jon        Snow  876-678-3469   
5         1006        Ron     Swanson  304-762-2467   
6         1007       Jeff      Winger                 
7         1008   Sherlock      Holmes  876-678-3469   
8         1009    Gandalf                             
9         1010      Peter      Parker  123-545-5421   
10        1011    Samwise      Gamgee                 
11        1012      Harry      Potter                 
12        1013        Don      Draper  123-543-2345   
13        1014     Leslie       Knope  876-678-3469   
14        1015       Toby  Flenderson  304-762-2467   
15        1016        Ron     Weasley  123-545-5421   
16        1017   Michael        Scott  123-643-9775   
17        1018      Clark        Kent                 
18        1019      Creed      Braton                 
19        1020     Anakin   Skywalker  876-678-3469   

                                  Address Paying Customer Do_Not_Contact  \
0                   123 Shire Lane, Shire               Y              N   
1                     93 West Main Street               N              Y   
2                      298 Drugs Driveway               N                  
3   980 Paper Avenue, Pennsylvania, 18503               Y              Y   
4                        123 Dragons Road               Y              N   
5                        768 City Parkway               Y              Y   
6                       1209 South Street               N              N   
7                           98 Clue Drive               N              N   
8                        123 Middle Earth               Y                  
9              25th Main Street, New York               Y              N   
10                  612 Shire Lane, Shire               Y              N   
11                   2394 Hogwarts Avenue               Y                  
12                       2039 Main Street               Y              N   
13                       343 City Parkway               Y              N   
14                          214 HR Avenue               N              N   
15                   2395 Hogwarts Avenue               N              N   
16         121 Paper Avenue, Pennsylvania               Y              N   
17                        3498 Super Lane               Y                  
18                                                    N/a              Y   
19            910 Tatooine Road, Tatooine               Y              N   

          Street_Address          State Zip_Code  
0         123 Shire Lane          Shire           
1    93 West Main Street                          
2     298 Drugs Driveway                          
3       980 Paper Avenue   Pennsylvania    18503  
4       123 Dragons Road                          
5       768 City Parkway                          
6      1209 South Street                          
7          98 Clue Drive                          
8       123 Middle Earth                          
9       25th Main Street       New York           
10        612 Shire Lane          Shire           
11  2394 Hogwarts Avenue                          
12      2039 Main Street                          
13      343 City Parkway                          
14         214 HR Avenue                          
15  2395 Hogwarts Avenue                          
16      121 Paper Avenue   Pennsylvania           
17       3498 Super Lane                          
18                   N/a                          
19     910 Tatooine Road       Tatooine

In [128]:
df["Paying Customer"] = df["Paying Customer"].str.replace("N/a",'')

df

CustomerID First_Name   Last_Name  Phone_Number  \
0         1001      Frodo     Baggins  123-545-5421   
1         1002       Abed       Nadir  123-643-9775   
2         1003     Walter       White                 
3         1004     Dwight     Schrute  123-543-2345   
4         1005        Jon        Snow  876-678-3469   
5         1006        Ron     Swanson  304-762-2467   
6         1007       Jeff      Winger                 
7         1008   Sherlock      Holmes  876-678-3469   
8         1009    Gandalf                             
9         1010      Peter      Parker  123-545-5421   
10        1011    Samwise      Gamgee                 
11        1012      Harry      Potter                 
12        1013        Don      Draper  123-543-2345   
13        1014     Leslie       Knope  876-678-3469   
14        1015       Toby  Flenderson  304-762-2467   
15        1016        Ron     Weasley  123-545-5421   
16        1017   Michael        Scott  123-643-9775   
17        1018      Clark        Kent                 
18        1019      Creed      Braton                 
19        1020     Anakin   Skywalker  876-678-3469   

                                  Address Paying Customer Do_Not_Contact  \
0                   123 Shire Lane, Shire               Y              N   
1                     93 West Main Street               N              Y   
2                      298 Drugs Driveway               N                  
3   980 Paper Avenue, Pennsylvania, 18503               Y              Y   
4                        123 Dragons Road               Y              N   
5                        768 City Parkway               Y              Y   
6                       1209 South Street               N              N   
7                           98 Clue Drive               N              N   
8                        123 Middle Earth               Y                  
9              25th Main Street, New York               Y              N   
10                  612 Shire Lane, Shire               Y              N   
11                   2394 Hogwarts Avenue               Y                  
12                       2039 Main Street               Y              N   
13                       343 City Parkway               Y              N   
14                          214 HR Avenue               N              N   
15                   2395 Hogwarts Avenue               N              N   
16         121 Paper Avenue, Pennsylvania               Y              N   
17                        3498 Super Lane               Y                  
18                                                                     Y   
19            910 Tatooine Road, Tatooine               Y              N   

          Street_Address          State Zip_Code  
0         123 Shire Lane          Shire           
1    93 West Main Street                          
2     298 Drugs Driveway                          
3       980 Paper Avenue   Pennsylvania    18503  
4       123 Dragons Road                          
5       768 City Parkway                          
6      1209 South Street                          
7          98 Clue Drive                          
8       123 Middle Earth                          
9       25th Main Street       New York           
10        612 Shire Lane          Shire           
11  2394 Hogwarts Avenue                          
12      2039 Main Street                          
13      343 City Parkway                          
14         214 HR Avenue                          
15  2395 Hogwarts Avenue                          
16      121 Paper Avenue   Pennsylvania           
17       3498 Super Lane                          
18                   N/a                          
19     910 Tatooine Road       Tatooine

# Drop those who said dont contact me

In [131]:
for x in df.index:
    if df.loc[x,"Do_Not_Contact"]=="Y":
        df.drop(x,inplace=True)

df

CustomerID First_Name   Last_Name  Phone_Number  \
0         1001      Frodo     Baggins  123-545-5421   
2         1003     Walter       White                 
4         1005        Jon        Snow  876-678-3469   
6         1007       Jeff      Winger                 
7         1008   Sherlock      Holmes  876-678-3469   
8         1009    Gandalf                             
9         1010      Peter      Parker  123-545-5421   
10        1011    Samwise      Gamgee                 
11        1012      Harry      Potter                 
12        1013        Don      Draper  123-543-2345   
13        1014     Leslie       Knope  876-678-3469   
14        1015       Toby  Flenderson  304-762-2467   
15        1016        Ron     Weasley  123-545-5421   
16        1017   Michael        Scott  123-643-9775   
17        1018      Clark        Kent                 
19        1020     Anakin   Skywalker  876-678-3469   

                           Address Paying Customer Do_Not_Contact  \
0            123 Shire Lane, Shire               Y              N   
2               298 Drugs Driveway               N                  
4                 123 Dragons Road               Y              N   
6                1209 South Street               N              N   
7                    98 Clue Drive               N              N   
8                 123 Middle Earth               Y                  
9       25th Main Street, New York               Y              N   
10           612 Shire Lane, Shire               Y              N   
11            2394 Hogwarts Avenue               Y                  
12                2039 Main Street               Y              N   
13                343 City Parkway               Y              N   
14                   214 HR Avenue               N              N   
15            2395 Hogwarts Avenue               N              N   
16  121 Paper Avenue, Pennsylvania               Y              N   
17                 3498 Super Lane               Y                  
19     910 Tatooine Road, Tatooine               Y              N   

          Street_Address          State Zip_Code  
0         123 Shire Lane          Shire           
2     298 Drugs Driveway                          
4       123 Dragons Road                          
6      1209 South Street                          
7          98 Clue Drive                          
8       123 Middle Earth                          
9       25th Main Street       New York           
10        612 Shire Lane          Shire           
11  2394 Hogwarts Avenue                          
12      2039 Main Street                          
13      343 City Parkway                          
14         214 HR Avenue                          
15  2395 Hogwarts Avenue                          
16      121 Paper Avenue   Pennsylvania           
17       3498 Super Lane                          
19     910 Tatooine Road       Tatooine

In [135]:
for x in df.index:
    if df.loc[x,"Phone_Number"]=='':
        df.drop(x,inplace=True)

df

CustomerID First_Name   Last_Name  Phone_Number  \
0         1001      Frodo     Baggins  123-545-5421   
4         1005        Jon        Snow  876-678-3469   
7         1008   Sherlock      Holmes  876-678-3469   
9         1010      Peter      Parker  123-545-5421   
12        1013        Don      Draper  123-543-2345   
13        1014     Leslie       Knope  876-678-3469   
14        1015       Toby  Flenderson  304-762-2467   
15        1016        Ron     Weasley  123-545-5421   
16        1017   Michael        Scott  123-643-9775   
19        1020     Anakin   Skywalker  876-678-3469   

                           Address Paying Customer Do_Not_Contact  \
0            123 Shire Lane, Shire               Y              N   
4                 123 Dragons Road               Y              N   
7                    98 Clue Drive               N              N   
9       25th Main Street, New York               Y              N   
12                2039 Main Street               Y              N   
13                343 City Parkway               Y              N   
14                   214 HR Avenue               N              N   
15            2395 Hogwarts Avenue               N              N   
16  121 Paper Avenue, Pennsylvania               Y              N   
19     910 Tatooine Road, Tatooine               Y              N   

          Street_Address          State Zip_Code  
0         123 Shire Lane          Shire           
4       123 Dragons Road                          
7          98 Clue Drive                          
9       25th Main Street       New York           
12      2039 Main Street                          
13      343 City Parkway                          
14         214 HR Avenue                          
15  2395 Hogwarts Avenue                          
16      121 Paper Avenue   Pennsylvania           
19     910 Tatooine Road       Tatooine

# Reset index (final dataset)

In [137]:
df = df.reset_index(drop=True)

df

CustomerID First_Name   Last_Name  Phone_Number  \
0        1001      Frodo     Baggins  123-545-5421   
1        1005        Jon        Snow  876-678-3469   
2        1008   Sherlock      Holmes  876-678-3469   
3        1010      Peter      Parker  123-545-5421   
4        1013        Don      Draper  123-543-2345   
5        1014     Leslie       Knope  876-678-3469   
6        1015       Toby  Flenderson  304-762-2467   
7        1016        Ron     Weasley  123-545-5421   
8        1017   Michael        Scott  123-643-9775   
9        1020     Anakin   Skywalker  876-678-3469   

                          Address Paying Customer Do_Not_Contact  \
0           123 Shire Lane, Shire               Y              N   
1                123 Dragons Road               Y              N   
2                   98 Clue Drive               N              N   
3      25th Main Street, New York               Y              N   
4                2039 Main Street               Y              N   
5                343 City Parkway               Y              N   
6                   214 HR Avenue               N              N   
7            2395 Hogwarts Avenue               N              N   
8  121 Paper Avenue, Pennsylvania               Y              N   
9     910 Tatooine Road, Tatooine               Y              N   

         Street_Address          State Zip_Code  
0        123 Shire Lane          Shire           
1      123 Dragons Road                          
2         98 Clue Drive                          
3      25th Main Street       New York           
4      2039 Main Street                          
5      343 City Parkway                          
6         214 HR Avenue                          
7  2395 Hogwarts Avenue                          
8      121 Paper Avenue   Pennsylvania           
9     910 Tatooine Road       Tatooine